### ISSUE TITLE
Warning when a variable is assigned a value incompatible with its enum type

### ISSUE THREAD

**Issue by sreed101** [palantir/tslint#4329]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**
General problem

**What does your suggested rule do?**
The rule would throw an error when a variable is assigned a value that is incompatible with its enum type.

**List several examples where your rule could be used**
```ts
export enum Gender {
    male = 1,
    female = 2
}

export class User {
    id: number;
    gender: Gender;
}

var user = new User();

user.gender = Gender.Male; // OK

user.gender = 3; // Currently is OK, desired result: TYPE ERROR

if (user.gender === Gender.Male) {} // OK

if (user.gender === 3) {} // Currently is OK, desired result: TYPE ERROR
```

**Additional context**
https://stackoverflow.com/questions/50846308/is-it-possible-to-have-tslint-type-check-enum-types?noredirect=1#comment93918362_50846308

**Comment by JoshuaKGoldberg** [comment]
Thanks for posting this @AdJesumPerMariam! Agreed that this would be useful. How exactly should the TSLint rule's logic go? There's a lot of info in the TypeScript issue, could you distill it here for the rule proposal?

Pulling in the original TypeScript feature request from the SO question: https://github.com/Microsoft/TypeScript/issues/26362

**Comment by chharvey** [comment]
So it looks like when the enum members are numbers, no error is thrown, whereas an error is (rightfully) thrown for enum members that are strings. See [this playground](http://www.typescriptlang.org/play/index.html#src=enum%20OrdinalPlain%20%7B%0D%0A%20%20%20%20FIRST%2C%0D%0A%20%20%20%20SECOND%2C%0D%0A%7D%0D%0Aenum%20OrdinalNumber%20%7B%0D%0A%20%20%20%20FIRST%20%3D%20100%2C%0D%0A%20%20%20%20SECOND%20%3D%20101%2C%0D%0A%7D%0D%0Aenum%20OrdinalString%20%7B%0D%0A%09FIRST%20%3D%20'first'%2C%0D%0A%09SECOND%20%3D%20'second'%2C%0D%0A%7D%0D%0A%0D%0Alet%20a%3A%20OrdinalPlain%20%3D%20OrdinalPlain.FIRST%0D%0Aa%20%3D%2042%20%2F%2F%20doesn%E2%80%99t%20throw%20an%20error%2C%20but%20is%20expected%20to.%0D%0A%0D%0Alet%20b%3A%20OrdinalNumber%20%3D%20OrdinalNumber.FIRST%0D%0Ab%20%3D%2042%20%2F%2F%20doesn%E2%80%99t%20throw%20an%20error%2C%20but%20is%20expected%20to.%0D%0A%0D%0Alet%20c%3A%20OrdinalString%20%3D%20OrdinalString.FIRST%0D%0Ac%20%3D%20'second'%20%2F%2F%20throws%20an%20error%20(good).%0D%0A%0D%0A%0D%0A) for a simple demo.

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
